collect/collect.go

```diff
@@ -287,22 +276,28 @@ func (i *InMemCollector) processSpan(sp *types.Span) {
 		// trace hasn't already been sent (or this span is really old); let's
 		// create a new trace to hold it
 		i.Metrics.IncrementCounter("trace_accepted")
-		cancel := make(chan struct{})
+
+		timeout, err := i.Config.GetTraceTimeout()
+
+		if err != nil {
+			// TODO: our implementation of config does not return errors
+			// maybe we should just remove the error from the signature
+			timeout = 60
+		}
+
 		trace = &types.Trace{
-			APIHost:       sp.APIHost,
-			APIKey:        sp.APIKey,
-			Dataset:       sp.Dataset,
-			TraceID:       sp.TraceID,
-			StartTime:     time.Now(),
-			CancelSending: cancel,
+			APIHost:   sp.APIHost,
+			APIKey:    sp.APIKey,
+			Dataset:   sp.Dataset,
+			TraceID:   sp.TraceID,
+			StartTime: time.Now(),
+			SendBy:    time.Now().Add(time.Duration(timeout) * time.Second),
 		}
 		// push this into the cache and if we eject an unsent trace, send it ASAP
 		ejectedTrace := i.Cache.Set(trace)
 		if ejectedTrace != nil {
-			i.sendASAP(ejectedTrace)
+			i.send(ejectedTrace)
```

Comment:
I'm curious to think more about what using a blocking send here means. Worth noting that `libhoney` is configured to `BlockOnSend`, so if transmission to Honeycomb slows down, that will push back up to here too.

I don't tihnk there is any pressure relief valve in the path between the public listener and processSpan.

The HTTP listener invokes the router that calls `r.Collector.AddSpan`
`r.Collector.AddSpan` attempts to push the event onto the `incoming` channel and blocks there.
The goroutine in `merge` is also a blocking write to the channel.
And that gets us to the `collect` loop's `select`, which calls `processSpan` in a synchronous manner.

I think the end result of this is that if Honeycomb slows down and libhoney can't send spans fast enough, that pressure will push all the way back to the main HTTP listener. Go will spin up goroutines per accepted incoming HTTP connection until it can't anymore and samproxy itself will become unavailable.

So the question becomes - should we release that pressure (probably?) and if so, where?

Maybe all the way up in the incoming router when we accept spans? If we change `r.Collector.AddSpan` to return an error when the channel is full, the HTTP listener could turn around and encode that as a 429 response or something else indicating that samproxy is overflowing, giving the client an opportunity to respond. 

I think I'd be ok splitting this idea into a different ticket / different PR, since the previous state would fail in a similar manner. (A little bit different - it would expand goroutine count until it tipped over within the delayed sends rather than within the HTTP listener, but same end result - death by goroutine.)

What do you think?